### PR TITLE
NEW @W-17015980@ Integrated --target functionality

### PR DESCRIPTION
**Pull request by jfeingold35**
This PR does the following:
- Replaces the hidden `--path-start` flag on `run.ts` with `--target`
- Adds `--target` to `rules` and `config`

### CODE REVIEW

File: messages/config-command.md
Comment (jfeingold35) [inline messages/config-command.md:1]: Sorta winged it for the message text on `config`.

File: messages/rules-command.md
Comment (jfeingold35) [inline messages/rules-command.md:1]: Sorta winged it for the message text on `rules`.

File: src/commands/code-analyzer/config.ts [hunk 1]
```diff
@@ -48,6 +55,9 @@ export default class ConfigCommand extends SfCommand<void> implements Displayabl
 
 	public async run(): Promise<void> {
 		const parsedFlags = (await this.parse(ConfigCommand)).flags;
+		if (parsedFlags.target && !parsedFlags.workspace) {
+			parsedFlags.workspace = ['.'];
```
Comment (jfeingold35) [inline src/commands/code-analyzer/config.ts:59]: In `config`, `--workspace` and `--target` are both optional, raising the possibility that `--target` could be provided but not `--workspace`.
Functionally, it doesn't really matter what we do in that case, but I made the decision to have the workspace default to the current directory.

File: src/commands/code-analyzer/rules.ts [hunk 2]
```diff
@@ -55,6 +62,9 @@ export default class RulesCommand extends SfCommand<void> implements Displayable
 
 	public async run(): Promise<void> {
 		const parsedFlags = (await this.parse(RulesCommand)).flags;
+		if (parsedFlags.target && !parsedFlags.workspace) {
+			parsedFlags.workspace = ['.'];
```
Comment (jfeingold35) [inline src/commands/code-analyzer/rules.ts:66]: In `rules`, `--workspace` and `--target` are both optional, raising the possibility that `--target` could be provided but not `--workspace`.
Functionally, it doesn't really matter what we do in that case, but I made the decision to have the workspace default to the current directory.

File: src/lib/utils/WorkspaceUtil.ts [hunk 3]
```diff
@@ -2,23 +2,29 @@ import * as fg from 'fast-glob';
 import {CodeAnalyzer, Workspace} from '@salesforce/code-analyzer-core';
 import {getMessage, BundleName} from '../messages';
 
-export async function createWorkspace(core: CodeAnalyzer, workspacePaths: string[]): Promise<Workspace> {
+export async function createWorkspace(core: CodeAnalyzer, workspacePaths: string[], targetPaths?: string[]): Promise<Workspace> {
```
Comment (jfeingold35) [inline src/lib/utils/WorkspaceUtil.ts:5]: This is probably the most consequential change. I'm validating `--target` the same way that I evaluate `--workspace`. I don't currently have any enforcement about stuff like, for example, making sure `--target` is a strict subset of `--workspace`, but I can add that if we need it.

File: test/commands/code-analyzer/config.test.ts
Comment (jfeingold35) [inline test/commands/code-analyzer/config.test.ts:1]: This file looks like it got changed a lot, but what happened is I rearranged some of the tests a little bit.

File: test/commands/code-analyzer/config.test.ts [hunk 4]
```diff
@@ -28,184 +28,188 @@ describe('`code-analyzer config` tests', () => {
...
 
-			it('Can be supplied once with multiple comma-separated values', async () => {
-				const inputValue =['./somedirectory', './someotherdirectory'];
-				await ConfigCommand.run(['--workspace', inputValue.join(',')]);
-				expect(executeSpy).toHaveBeenCalled();
-				expect(receivedActionInput).toHaveProperty('workspace', inputValue);
-			});
+	describe.each([
+		{flag: '--workspace', shortflag: '-w', property: 'workspace'},
+		{flag: '--target', shortflag: '-t', property: 'target'}
```
Comment (jfeingold35) [inline test/commands/code-analyzer/config.test.ts:115]: Since `--target` and `--workspace` share all the same constraints and expectations on a flag-level, I just made them variants of the same `describe.each` call.

File: test/commands/code-analyzer/config.test.ts [hunk 5]
```diff
@@ -28,184 +28,188 @@ describe('`code-analyzer config` tests', () => {
...
+			expect(executeSpy).toHaveBeenCalled();
+			expect(createActionSpy).toHaveBeenCalled();
+			expect(fromFileSpy).not.toHaveBeenCalled();
+			expect(receivedActionInput['output-file']).toBeUndefined();
+		});
+	});
+
+	describe('Flag interactions', () => {
+		describe('--workspace and --target', () => {
+			it('If --target is specified, --workspace defaults to "."', async () => {
```
Comment (jfeingold35) [inline test/commands/code-analyzer/config.test.ts:206]: As mentioned, testing for the edge case where `--target` and not `--workspace` is specified.

File: test/commands/code-analyzer/rules.test.ts [hunk 6]
```diff
@@ -200,35 +200,38 @@ describe('`code-analyzer rules` tests', () => {
 		});
 	});
 
-	describe('--workspace', () => {
+	describe.each([
+		{flag: '--workspace', shortflag: '-w', property: 'workspace'},
+		{flag: '--target', shortflag: '-t', property: 'target'},
```
Comment (jfeingold35) [inline test/commands/code-analyzer/rules.test.ts:205]: Since `--target` and `--workspace` share all the same constraints and expectations on a flag-level, I just made them variants of the same `describe.each` call.

File: test/commands/code-analyzer/rules.test.ts [hunk 7]
```diff
@@ -237,15 +240,25 @@ describe('`code-analyzer rules` tests', () => {
...
-			await RulesCommand.run(['-w', inputValue]);
+			await RulesCommand.run([shortflag, inputValue]);
 			expect(executeSpy).toHaveBeenCalled();
-			expect(receivedActionInput).toHaveProperty('workspace', [inputValue]);
+			expect(receivedActionInput).toHaveProperty(property, [inputValue]);
 		});
 	});
 
 	describe('Flag interactions', () => {
+		describe('--workspace and --target', () => {
```
Comment (jfeingold35) [inline test/commands/code-analyzer/rules.test.ts:252]: Validation for case where `--target` is used and `--workspace` isn't.

File: test/commands/code-analyzer/run.test.ts [hunk 8]
```diff
@@ -80,42 +80,42 @@ describe('`code-analyzer run` tests', () => {
 		});
 	});
 
-	describe('--path-start', () => {
+	describe('--target', () => {
```
Comment (jfeingold35) [inline test/commands/code-analyzer/run.test.ts:83]: Straightforward conversion of `--path-start` tests into `--target` tests.

File: test/lib/actions/ConfigAction.test.ts [hunk 9]
```diff
@@ -398,6 +398,41 @@ describe('ConfigAction tests', () => {
 		});
 	});
 
+	describe('Target/Workspace resolution', () => {
```
Comment (jfeingold35) [inline test/lib/actions/ConfigAction.test.ts:401]: Validates both the `--workspace` flag in isolation and the `--workspace` flag in conjunction with `--target`.

File: src/commands/code-analyzer/config.ts (hunk 1 above)
Comment (stephen-carter-at-sf) [reply src/commands/code-analyzer/config.ts:59]: In the proposal document, it says:

```
If the --workspace flag is not specified then:
  - if one or more files are explicitly targeted via the --target flag, then 
    the targeted files will be used as the workspace.
  - if a --target flag is also not specified, then the present working folder 
    (i.e. '.') will be used and all its files will be targeted.
```

So in this case the workspace should equal the target files

**Review by stephen-carter-at-sf — CHANGES_REQUESTED** [review]

File: src/commands/code-analyzer/config.ts (hunk 1 above)
Comment (stephen-carter-at-sf) [reply src/commands/code-analyzer/config.ts:59]: Discussed offline and we decided that Josh is right here - but we should probably move this logic to the action layer instead of the command layer.

File: messages/config-command.md [hunk 10]
```diff
@@ -46,35 +46,37 @@ Set of files that make up your workspace.
...
+Use the `--workspace` flag to display only the configuration associated with the rules that apply to the files that make up your workspace. Typically, a workspace is a single project folder that contains all your files. But it can also consist of one or more folders, one or more files, and use glob patterns (wildcards). If you specify this flag multiple times, then your workspace is the sum of the files and folders.
 
-This command uses the type of file in the workspace, such as JavaScript or Typescript, to determine the rules to include in the configuration state. For example, if your workspace contains only JavaScript files, the command doesn't include TypeScript rules. The command uses a file's extension to determine what kind of file it is, such as ".ts" for TypeScript.
+This command uses the type of file in the workspace, such as JavaScript or Typescript, to determine which configuration state is applicable. For example, if your workspace contains only JavaScript files,  then the command doesn't display configuration state associated with TypeScript rules. The command uses a file's extension to determine what kind of file it is, such as ".ts" for TypeScript.
```
Comment (jfeingold35) [inline messages/config-command.md:51]: This should probably be `types of file`, right?

File: messages/config-command.md [hunk 11]
```diff
@@ -42,25 +42,41 @@ We're continually improving Salesforce Code Analyzer. Tell us what you think! Gi
...
+Use the `--workspace` flag to display only the configuration associated with the rules that apply to the files that make up your workspace. Typically, a workspace is a single project folder that contains all your files. But it can also consist of one or more folders, one or more files, and use glob patterns (wildcards). If you specify this flag multiple times, then your workspace is the sum of the files and folders.
 
-This command uses the type of file in the workspace, such as JavaScript or Typescript, to determine the rules to include in the configuration state. For example, if your workspace contains only JavaScript files, the command doesn't include TypeScript rules. The command uses a file's extension to determine what kind of file it is, such as ".ts" for TypeScript.
+This command uses the types of file in the workspace, such as JavaScript or Typescript, to determine which configuration state is applicable. For example, if your workspace contains only JavaScript files,  then the command doesn't display configuration state associated with TypeScript rules. The command uses a file's extension to determine what kind of file it is, such as ".ts" for TypeScript.
```
Comment (stephen-carter-at-sf) [inline messages/config-command.md:51]: Small little thing: Looks like there is 2 spaces between the words: `files,  then`

Comment (jfeingold35) [reply messages/config-command.md]: Fixed. Also removed some other double-spaces.

File: src/commands/code-analyzer/rules.ts [hunk 12]
```diff
@@ -55,6 +62,9 @@ export default class RulesCommand extends SfCommand<void> implements Displayable
...
+		}
```
Comment (stephen-carter-at-sf) [inline src/commands/code-analyzer/rules.ts:67]: Can we move this to the action layer?

File: src/commands/code-analyzer/config.ts [hunk 13]
```diff
@@ -48,6 +55,9 @@ export default class ConfigCommand extends SfCommand<void> implements Displayabl
...
+		}
```
Comment (stephen-carter-at-sf) [inline src/commands/code-analyzer/config.ts:60]: Can we move this to the action layer?

File: src/lib/utils/WorkspaceUtil.ts (hunk 3 above)
Comment (stephen-carter-at-sf) [reply src/lib/utils/WorkspaceUtil.ts:5]: The core side is already validating that target is underneath workspace. We just need the cli to expand globs if needed and pass the files through.

See what happens when you interactively supply --target outside of the pwd and see what the error looks like.

File: test/lib/actions/ConfigAction.test.ts [hunk 14]
```diff
@@ -399,18 +399,32 @@ describe('ConfigAction tests', () => {
...
+		});
+
+		afterEach(() => {
+			process.chdir(originalCwd);
+		});
+
 		it.each([
 			{
-				case: 'a workspace is applied to the config',
+				case: 'an workspace is applied to the config',
```
Comment (stephen-carter-at-sf) [inline test/lib/actions/ConfigAction.test.ts:415]: typo: "an workspace" --> "a workspace"

**Review by stephen-carter-at-sf — APPROVED** [review]

File: messages/run-command.md [hunk 15]
```diff
@@ -52,45 +52,41 @@ We're continually improving Salesforce Code Analyzer. Tell us what you think! Gi
...
 
-Set of files you want to include in the code analysis. 
+Set of files that make up your workspace.
 
 # flags.workspace.description
 
-Typically, a workspace is a single project folder that contains all your files. But it can also consist of one or more folders, one or more files, and use glob patterns (wildcards). If you specify this flag multiple times, then your workspace is the sum of the files and folders. 
+Typically, a workspace is a single project folder that contains all your files. But it can also consist of one or more folders, one or more files, and use glob patterns (wildcards). If you specify this flag multiple times, then your workspace is the sum of the files and folders.
+
+Note that even if you wish to only target (via the `--target` flag) a subset of the files within your workspace, some engines may still require your complete workspace to perform its analysis on your targeted files. For example, the Salesforce Graph Engine may need to compile your entire project in order to properly build a graph so that it may perform data flow analysis on the paths that start in your targeted files.
```
Comment (jshackell-sfdc) [inline messages/run-command.md:63]:
```suggestion
Some engines often need your entire code base to perform an analysis, even if you want to target only a subset of the files within your workspace , such as with the `--target` flag. For example, the Salesforce Graph Engine might need to compile your entire project in order to properly build a graph so it can perform a data flow analysis on the paths that start in your targeted files.
```

File: messages/run-command.md [hunk 16]
```diff
@@ -52,45 +52,41 @@ We're continually improving Salesforce Code Analyzer. Tell us what you think! Gi
...
 
-# flags.path-start.summary
+If the `--workspace` flag is not specified then the current folder '.' will be used.
```
Comment (jshackell-sfdc) [inline messages/run-command.md:65]:
```suggestion
If you don't specify the `--workspace` flag, then the current folder '.' is used as your workspace.
```

File: messages/run-command.md [hunk 17]
```diff
@@ -52,45 +52,41 @@ We're continually improving Salesforce Code Analyzer. Tell us what you think! Gi
...
+# flags.target.summary
 
-# flags.path-start.description
+Subset of files within your workspace to be targeted for analysis.
 
-If you don't specify this flag, then any path-based analysis rules automatically discover and use all starting points found in your workspace. Use this flag to restrict the starting points to only those you want in your code analysis. 
+# flags.target.description
 
-This flag only applies to path-based analysis rules, which are of type DataFlow and Flow. These types of rules are only available from some engines, like the Salesforce Graph Engine, "sfge" for example.
+A target may be specified as a file, a folder, or a glob pattern.
```
Comment (jshackell-sfdc) [inline messages/run-command.md:73]:
```suggestion
You can specify a target as a file, a folder, or a glob pattern.
```

File: messages/run-command.md [hunk 18]
```diff
@@ -52,45 +52,41 @@ We're continually improving Salesforce Code Analyzer. Tell us what you think! Gi
...
 
-If you specify a file or a folder as your starting point, then the analysis uses only the methods that have public or global accessibility. 
+If you specify this flag multiple times, then the full list of targeted files will be the sum of the files and folders.
```
Comment (jshackell-sfdc) [inline messages/run-command.md:75]:
```suggestion
If you specify this flag multiple times, then the full list of targeted files is the sum of the files and folders.
```

File: messages/run-command.md [hunk 19]
```diff
@@ -52,45 +52,41 @@ We're continually improving Salesforce Code Analyzer. Tell us what you think! Gi
...
 
-To specify individual methods as a starting point, use the syntax "<file>#methodName" to select a single method or "<file>#methodName1;methodName2" to select multiple methods. For example, "SomeClass.cls#method1" (single method) or "SomeClass.cls#method1;method2" (multiple methods).
+Each targeted file must live within the workspace specified by the –-workspace flag.
```
Comment (jshackell-sfdc) [inline messages/run-command.md:77]:
```suggestion
Each targeted file must live within the workspace that you specified with the `–-workspace` flag.
```

File: messages/run-command.md [hunk 20]
```diff
@@ -52,45 +52,41 @@ We're continually improving Salesforce Code Analyzer. Tell us what you think! Gi
...
 
-You can use glob patterns (wildcards) only when specifying files and folders; you can't use glob patterns when specifying individual methods.
+If the `--target` flag is not specified, then all the files within your workspace (supplied by the `--workspace` flag) will be targeted for analysis.
```
Comment (jshackell-sfdc) [inline messages/run-command.md:79]:
```suggestion
If you don't specify the `--target` flag, then all the files within your workspace (specified by the `--workspace` flag) are targeted for analysis.
```

**Review by jshackell-sfdc — CHANGES_REQUESTED** [review]
@jfeingold35 , here's my edit.  I did my best to catch everything, but let me know if I got something wrong.

**Review by jshackell-sfdc — APPROVED** [review]
